tools/privacyidea-migrate-linotp.py

```diff
@@ -1,84 +1,161 @@
 privacyIDEA with the resolver "PIResolver" in realm "pirealm".
 
-Tokens, that were not assigned to a user will be assigned to the realms "realm1" and "realm2"
-in privacyIDEA.
-"""
-ASSIGNMENTS = {
-    "resolver": {"lokal": "PIResolver"},
-    "realm": {"PIResolver": "pirealm"},
-    "unassigned_tokens": ["realm1", "realm2"]
-}
+Tokens, that were not assigned to a user will be assigned to the realms specified
+in the list "unassigned_tokens".
 
+The NEW_TOKENINFO section can set any arbitrary tokeninfo for the 
+migrated tokens.
 """
-This option splits INSERT statements into chunks of a given size, which is useful
-if the number of migrated tokens is so large that it cannot fit in one INSERT
-statement.
-"""
-INSERT_CHUNK_SIZE = 10000
-
-# Do not change anything after this line
-# ============================================================================
-#
-
-
-def migrate():
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.sql import select
+from sqlalchemy.schema import ForeignKey
+from sqlalchemy import (Table, MetaData, Column, Integer, Unicode, Boolean,
+                        UnicodeText)
+import sys
+import json
+import getopt
+import binascii
+import os
+import re
+from privacyidea.lib.crypto import aes_cbc_decrypt, aes_cbc_encrypt
+from privacyidea.lib.utils import hexlify_and_unicode
+
+
+EXAMPLE_CONFIG_FILE = """{
+    "SQL": {
+        "LINOTP_URI": "mysql+pymysql://linotp:linotp@localhost/linotp2",
+        "PRIVACYIDEA_URI": "mysql+pymysql://pi:QEK_k_f3SGbu@localhost/pi?charset=utf8mb4",
+        "INSERT_CHUNK_SIZE": 10000
+    },
+    "MIGRATE": {
+        "tokens": true,
+        "tokeninfo": true,
+        "assignments": false
+    },
+    "ASSIGNMENTS": {
+        "resolver": {"local": "localpw"},
+        "realm": {"localpw": "neuerrealm"},
+        "unassigned_tokens": ["testfoo"],
+        "convert_endian": true
+    },
+    "ENCRYPTION": {
+        "reencrypt": false,
+        "linotp_enc_file": "/etc/linotp2/encKey",
+        "privacyidea_enc_file": "/etc/privacyidea/enckey"
+    },
+    "NEW_TOKENINFO": {
+        "arbitrary_migrationkey": "somevalue"
+    }
+}"""
+
+
+# These patterns are used to convert a middle endian UUID to a standard UUID:
+# 33221100554477668899AABBCCDDEEFF -> 00112233-4455-6677-8899-AABBCCDDEEFF
+UUID_MATCH_PATTERN = "^([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})" \
+                     "([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})" \
+                     "([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})" \
+                     "([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})([0-9a-fA-F]{2})$"
+
+UUID_REPLACE_PATTERN = "\\4\\3\\2\\1-\\6\\5-\\8\\7-\\9\\10-\\11\\12\\13\\14\\15\\16"
+
+
+class Config(object):
+
+    def __init__(self, config_file):
+        with open(config_file, "r") as f:
+            contents = f.read()
+        config = json.loads(contents)
```

Comment:
should we check if the file is really readable and in json format (with try ... except)?

Reply:
s.o. If the file is not readable we get a well readable exception, which is usually more readable to me than an error message: "File can not be read".
Then I wonder "which file"?

Usually the user, who runs the script also created the config file.
So I do not see the necessitiy to do excessive error handling.

Reply:
ok